pilot/test/test_esstager.py

```diff
@@ -0,0 +1,314 @@
+
+            xdata = [new_file_spec]
+            workdir = os.path.dirname(output_file)
+            client = StageInESClient(infoservice)
+            kwargs = dict(workdir=workdir, cwd=workdir, usecontainer=False)
+            client.transfer(xdata, activity=['es_events_read'], **kwargs)
+        except exception.PilotException, error:
+            logger.error("Pilot Exeception: %s, %s" % (error.get_detail(), traceback.format_exc()))
+        except Exception, e:
+            logger.error(traceback.format_exc())
+            error = exception.StageInFailure("stagein failed with error=%s" % e)
+
+        logger.info('Summary of transferred files:')
+        for e in xdata:
+            logger.info(" -- lfn=%s, status_code=%s, status=%s" % (e.lfn, e.status_code, e.status))
+
+        if error:
+            logger.error('Failed to stage-in eventservice file(%s): error=%s' % (output_file, error.get_detail()))
+            raise error
+
+    @unittest.skipIf(not check_env(), "No CVMFS")
+    def test_stageout_noexist_activity_stagein(self):
+        """
+        Make sure that no exceptions to stage out file.
+        """
+        error = None
+        try:
+            from pilot.info import infosys, InfoService
+            infoservice = InfoService()
+            infoservice.init('BNL_CLOUD_MCORE', infosys.confinfo, infosys.extinfo)
+
+            output_file = os.path.join('/tmp', str(uuid.uuid4()))
+            shutil.copy('/bin/hostname', output_file)
+            file_data = {'scope': 'transient',
+                         'lfn': os.path.basename(output_file),
+                         #'ddmendpoint': None,
+                         #'type': 'es_events',
+                         #'surl': output_file
+                         #'turl': None,
+                         #'filesize': None,
+                         #'checksum': None
+                         }
+            file_spec = FileSpec(type='output', **file_data)
+            xdata = [file_spec]
+            workdir = os.path.dirname(output_file)
+            client = StageOutESClient(infoservice)
+            kwargs = dict(workdir=workdir, cwd=workdir, usecontainer=False)
+            client.transfer(xdata, activity=['es_events_no_exist', 'pw'], **kwargs)
+        except exception.PilotException, error:
+            logger.error("Pilot Exeception: %s, %s" % (error.get_detail(), traceback.format_exc()))
+        except Exception, e:
+            logger.error(traceback.format_exc())
+            error = exception.StageOutFailure("stageOut failed with error=%s" % e)
+
+        logger.info('Summary of transferred files:')
+        for e in xdata:
+            logger.info(" -- lfn=%s, status_code=%s, status=%s" % (e.lfn, e.status_code, e.status))
+
+        if error:
+            logger.error('Failed to stage-out eventservice file(%s): error=%s' % (output_file, error.get_detail()))
+            raise error
+
+        storage_id = storage_maps.get_storage_id(file_spec.ddmendpoint)
```

Comment:
Did not you the the confusion logic in workflow:

1. you resolve `strorage_id` from `ddmendpoint` and pass it into the compound `storage_token`
2. and later you will parse back `storage_id` from `storage_token` and lookup back `ddmendpoint` name:)

hand-made complication, no?

Reply:
It's a test function.
The main purpose of this part is to test that it can stagein a file with scope, name and storage_token. It's what we have for esmerge jobs. But to make sure that a file is available at remote storage, I need to transfer a file to remote storage at first. I can do it with a setup function. It's another option. Here I at first transfer the file to remote storage and get the storage_id. Then I tested whether I can get it back.

Reply:
In this test function, the part before line 291 is just preparing the file.